assets/sip-big_vector/big_vector.move

```diff
@@ -0,0 +1,339 @@
+        BigVector<Element> {
+            id,
+            slice_count,
+            slice_size,
+            length: 0,
+        }
+    }
+
+    /// return the slice_count of the BigVector
+    public fun slice_count<Element: store>(bv: &BigVector<Element>): u64 {
+        bv.slice_count
+    }
+
+    /// return the max size of a slice in the BigVector
+    public fun slice_size<Element: store>(bv: &BigVector<Element>): u64 {
+        bv.slice_size
+    }
+
+    /// return the size of the BigVector
+    public fun length<Element: store>(bv: &BigVector<Element>): u64 {
+        bv.length
+    }
+
+    /// return the slice_id related to the index i
+    public fun slice_id<Element: store>(bv: &BigVector<Element>, i: u64): u64 {
+        (i / bv.slice_size) + 1
+    }
+
+    /// return true if the BigVector is empty
+    public fun is_empty<Element: store>(bv: &BigVector<Element>): bool {
+        bv.length == 0
+    }
+
+    /// push a new element at the end of the BigVector
+    public fun push_back<Element: store>(bv: &mut BigVector<Element>, element: Element) {
+        if (length(bv) / bv.slice_size == bv.slice_count) {
+            bv.slice_count = bv.slice_count + 1;
+            let new_slice = vector::singleton(element);
+            dynamic_field::add(&mut bv.id, bv.slice_count, new_slice);
+        }
+        else {
+            let slice = dynamic_field::borrow_mut(&mut bv.id, bv.slice_count);
+            vector::push_back(slice, element);
+        };
+        bv.length = bv.length + 1;
+    }
+
+    /// pop an element from the end of the BigVector
+    public fun pop_back<Element: store>(bv: &mut BigVector<Element>): Element {
+        let slice = dynamic_field::borrow_mut(&mut bv.id, bv.slice_count);
+        let element = vector::pop_back(slice);
+        trim_slice(bv);
+        bv.length = bv.length - 1;
+
+        element
+    }
+
+    /// borrow an element at index i from the BigVector
+    public fun borrow<Element: store>(bv: &BigVector<Element>, i: u64): &Element {
+        let slice_count = (i / bv.slice_size) + 1;
+        let slice = dynamic_field::borrow(&bv.id, slice_count);
+        vector::borrow(slice, i % bv.slice_size)
+    }
+
+    /// borrow a mutable element at index i from the BigVector
+    public fun borrow_mut<Element: store>(bv: &mut BigVector<Element>, i: u64): &mut Element {
+        let slice_count = (i / bv.slice_size) + 1;
+        let slice = dynamic_field::borrow_mut(&mut bv.id, slice_count);
+        vector::borrow_mut(slice, i % bv.slice_size)
+    }
+
+    /// borrow a slice from the BigVector
+    public fun borrow_slice<Element: store>(bv: &BigVector<Element>, slice_count: u64): &vector<Element> {
+        dynamic_field::borrow(&bv.id, slice_count)
+    }
+
+    /// borrow a mutable slice from the BigVector
+    public fun borrow_slice_mut<Element: store>(bv: &mut BigVector<Element>, slice_count: u64): &mut vector<Element> {
```

Comment:
this API is problematic because it breaks internal invariants of the `big_vector`.
As a reference consider the following:
```        
  // create a big vector of length 10 and slice size 3 as in your test ...
  assert!(big_vector::length(&big_vector) == 10, 100);
  let slice = big_vector::borrow_slice_mut(&mut big_vector, 1);
  vector::remove(slice, 2); // this breaks internal invariants of big_vector
  assert!(big_vector::length(&big_vector) == 10, 100); // the length is still the same 
  // as an example this call now breaks
  big_vector::remove(&mut big_vector, 2);
```
moreover in some scenario elements in a slice are not accessible any longer if in the code above I push something in the slice instead of removing. The slice will now have elements that I cannot access with the big vector API, or with any assumption over slice size, with all implications of that.

I have not explored all implications but I suspect that may break the code very seriously.

Thoughts?

Reply:
Since gas fees for dynamic_field accesses are charged by the number of function calls according to the [protocol config](https://github.com/MystenLabs/sui/blob/main/crates/sui-protocol-config/src/lib.rs#L1249), theses functions seems to be necessary and the developers/auditors may need to use these functions carefully.
Let me give you an [example](https://github.com/Typus-Lab/typus-framework/blob/main/compare-big-vector-borrow-gas-fee/sources/test.move), and here are the transactions of the two entry functions: [test_1](https://testnet.suivision.xyz/txblock/4Yua2TdNmcybvwwCobaGg9eSFSAuZxpJ4K7BKFquNSZG), [test_2](https://testnet.suivision.xyz/txblock/CirLeSXWntm144rLKWhSUKPqE4usHbot3B8a2AoA6zgv). When you dig into the gas fee section, you can find the significant computing fee difference. If the gas fee calculation or the natvie dynamic_field borrowing can be more optimize, I believe these two slice borrowing functions can be removed. This is the reason why we exposed these functions with risks if the developers use it in a wrong way.
However, I think there's still a chance to avoid your concern such as wrapping the std::vector into a struct and expose only the necessary iteration functions and the functions with risks can be friend functions. How do you think about this solution?

Reply:
I can see how in an iteration scenario having the slice will help. Any alternative would load the dynamic field over and over again. This is interesting.
In general to me the `mut` API is a non starter, any operation that affects the size is going to break the big vector, that is not an option.
As for wrapping the vector that is a super tricky thing. You'd have to move it out of the dynamic field which means the big vector would be "crippled" until the vector is going back.
I cannot think of a simple way to do this without some kind of redesign which I am not sure we want to do.
Happy to hear ideas though, as I may be missing something and feeling too doomed at the moment.
Also tagging people that may have some good idea on the subject @sblackshear @damirka @amnn @tnowacki @tzakian

Reply:
My bias is towards the wrapped vector approach, where the slice is taken, and put in a  hot potato that can give mutable access to it, and other operations that need to access slices could fail because they are accessing a taken slice.  It's not that I'm entirely happy with this solution, but it's the last bad one in my mind.

Reply:
Actually I take this back -- you don't need the hot potato approach, you can wrap the vector in another type as it sits in the the dynamic field, offer an API for borrowing mutable versions of that type, and an API on that type that only allows you to mutably borrow the underlying elements, without altering the dimensions of the vector:

```rust
struct Slice<E> has store {
  elements: vector<E>
}

public fun borrow<E: store>(self: &Slice<E>, ix: u64): &E {}

public fun borrow_mut<E: store>(self: &mut Slice<E>, ix: u64): &mut E {}

public fun borrow_slice<E: store>(self: &BigVector<E>, sx: u32): &Slice<E> {}

public fun borrow_slice_mut<E: store>(self: &mut BigVector<E>, sx: u32): &mut Slice<E> {}
```

Reply:
another option is to look into the gas charges a bit more too, I am realizing that maybe the way we are charging here may not be correct. I want to study the example a bit more and see if there is a solution that does not require any special solution. As @sblackshear pointed out the principal should be that you work with this as you would do with any array

Reply:
as I go and look into the gas story one could claim that taking the slice **AND** invalidating any operation on the big vector is the runtime equivalent of what would happen if the struct could hold a mutable reference to the vector. 
When/if we support that, the big vector would be unusable while you hold the slice.
And as disgusting as the runtime solution is, it is a somewhat-quasi-almost-kind-of similar...

Reply:
also I think (and correct me if I am wrong) the hot potato pattern could force the user to return the slice to the big vector, which would be nice. Is that right?

Reply:
I think the wrapping solution may be better if we don't want to let the developers call vector::remove, vector::add to the slice, and we won't need complex assertions to check if the returned slice is compromised or not. These slice borrowing functions are aimed to reduce the computation gas fee, if the gas charging can be optimized then these functions can be removed.